taichi/backends/metal/codegen_metal.cpp

```diff
@@ -748,17 +736,15 @@ class KernelCodegen : public IRVisitor {
         "1);",
         kRuntimeVarName);
     emit("device ListManager *parent_list = &({}->snode_lists[{}]);",
-         kRuntimeVarName, stmt->snode->parent->id);
-    emit("device ListManager *child_list = &({}->snode_lists[{}]);",
          kRuntimeVarName, sn_id);
-    emit("const SNodeMeta child_meta = {}->snode_metas[{}];", kRuntimeVarName,
+    emit("const SNodeMeta parent_meta = {}->snode_metas[{}];", kRuntimeVarName,
          sn_id);
-    emit("const int child_stride = child_meta.element_stride;");
-    emit("const int child_num_slots = child_meta.num_slots;");
+    emit("const int child_stride = parent_meta.element_stride;");
+    emit("const int child_num_slots = parent_meta.num_slots;");
     // Grid-stride loops:
     // Each thread begins at thread_index, and incremets by grid_size
-    emit("for (int ii = {}; ii < child_list->max_num_elems; ii += {}) {{",
-         kKernelThreadIdName, kKernelGridSizeName);
+    emit("for (int ii = {}; ii < {}; ii += {}) {{", kKernelThreadIdName,
+         total_num_elems_from_root, kKernelGridSizeName);
```

Comment:
Is the upper bound `total_num_elems_from_root` necessary?
We already have
`if (parent_idx_ >= num_active(parent_list)) return;` below.

Reply:
Actually, this is the struct-for kernel, and it is iterating over the **cells** (children) of the last `listgen Sx`. And in this context, parent refers to `Sx` itself.

Still using this example

```py
block = ti.root.pointer(ti.i, 4)
X = block.dense(ti.i, 8).place(x)
```

This iterates over the 4x8=32 cells of `X`..

Reply:
Thanks for the details! I'm still a little confused. If I understand correctly, by the time `ii >= total_num_elems_from_root`, we always have `parent_idx_ = ii / child_num_slots >= num_active(parent_list)`? So why do we need to bounds here, given one is always tighter than the other?

Reply:
Ah I (maybe) get your point now. You are suggesting to just remove the upper bound, e.g.

```cpp
for (int ii = thread_id;; ii += grid_size) {
  const int parent_idx_ = (ii / child_num_slots);
  if (parent_idx_ >= num_active(parent_list)) return;
  ...
}
```
? Good point, maybe I can remove the upper bound in the listgen kernel as well..

Reply:
Hmmm, I think this cannot be removed after some testing. This is because `total_num_elems_from_root` is **not** padded to power of two, while `child_num_slots` **is**. So if we have a settings whose leaf size is a) not POT, and b) over 65536 (the max #threads for listgen/struct-for kernels), then we need this additional guard to make sure the loop terminates precisely at `total_num_elems_from_root`.

This test case actually catches this case when I removed the upper bound in the **listgen** kernels (not struct-for, but I think the principles are similar):

https://github.com/taichi-dev/taichi/blob/d5818e58722053dbaa28e85966b330bd3a32d0ca/tests/python/test_bitmasked.py#L125-L128

Reply:
> maybe it makes sense to represent list generation as a struct-for

Hmm, did I actually say that before.. 🤔

Reply:
Sorry, I probably 脑补 the rest based on your proposal on finer granularity IR: https://github.com/taichi-dev/taichi/issues/562#issuecomment-595456800 

:-)